Patterns: Add user categories to site editor sidebar navigation screen

glendaviesnz opened the pull request:
## What?
Lists user patterns by category in the site editor patterns UI.

Depends on [User Patterns Category Select PR](https://github.com/WordPress/gutenberg/pull/53835).

## Why?
With the [addition the ability for users to categorise their patterns](https://github.com/WordPress/gutenberg/pull/53835) we also need a way for them to be displayed by category in the site editor patterns page.

## How?
Adds the user added categories to the category listing in left sidebar of site editor patterns page.

This PR was part of a group of PRs that were merged back into https://github.com/WordPress/gutenberg/pull/53835 before being merged into trunk.

A full list of the PRs involved in this work can be [found here](https://github.com/WordPress/gutenberg/issues/53164).

## Testing Instructions
 - Add several user patterns, synced and unsynced, with categories assigned (make sure to hit enter or add ',' after each category - this is the default tag input component behaviour - we will fix this in a follow up PR)
 - Go to the site editor patterns page and make sure the categories appear in the left nav panel
 - Make sure that when the categories are selected the correct patterns display
 - Make sure the theme patterns also still display as expected
 - Make sure that all the patterns display under 'All patterns' and that searching works as expected
 - Duplicate a theme pattern and make sure the new pattern is in the same category
 - Add a user pattern with category of `Footer` and make sure it displays along with the theme Footer patterns (be aware that although the TT3 theme footer patterns are listed under `Footers` the category slug is actually `footer` so the user category name needs to match this).


## Screenshots or screencast <!-- if applicable -->

https://github.com/WordPress/gutenberg/assets/3629020/736bf50d-f8d0-4a31-91f8-31ecb132d4b3

glendaviesnz committed:
Add pattern categories to site editor pattern listing

glendaviesnz committed:
Get the user patterns merged in with theme patterns

glendaviesnz committed:
Account for uncategorised patterns

aaronrobertshaw commented on packages/edit-site/src/components/page-patterns/use-patterns.js:
```diff
@@ -125,32 +133,48 @@ const selectThemePatterns = ( select, { categoryId, search = '' } = {} ) => {
+	name: patternBlock.slug,
+	syncStatus: patternBlock.wp_pattern_sync_status || SYNC_TYPES.full,
+	title: patternBlock.title.raw,
+	type: 'wp_block',
```
Do we need to hardcode this?

glendaviesnz replied:
no, fixed.

aaronrobertshaw commented on packages/edit-site/src/components/page-patterns/use-patterns.js:
```diff
@@ -125,32 +133,48 @@ const selectThemePatterns = ( select, { categoryId, search = '' } = {} ) => {
 
 	let patterns = records
-		? records.map( ( record ) => reusableBlockToPattern( record ) )
+		? records.map( ( record ) =>
+				patternBlockToPattern( record, categories )
+		  )
```
Could we create a map of the categories for [easier lookup](https://github.com/WordPress/gutenberg/pull/53837/files#diff-0f49c90dc39786ecdf45f374247e433f763d295d80bdaecda3bab11357e4d7fdR147-R149) when converting to a pattern object? Might make a difference when sites have 15K patterns and start to categorise them all.

glendaviesnz replied:
Done.

aaronrobertshaw commented on packages/patterns/src/components/category-selector.js:
```diff
@@ -0,0 +1,113 @@
+	function onChange( termNames ) {
+		const uniqueTerms = termNames.reduce( ( acc, name ) => {
+			if (
+				! acc.some( ( n ) => n.toLowerCase() === name.toLowerCase() )
```
Nit: Might make grokking this a little smoother if we avoid the abstract variable name. Similar [feedback](https://github.com/WordPress/gutenberg/pull/51078#discussion_r1236233036) was left on the original patterns work.

glendaviesnz replied:
I fixed this [over on the parent PR](https://github.com/WordPress/gutenberg/pull/53835/commits/0b7df8db9086b5ce4f60b38013e0a8f3753663bb) - file was only appearing here due top a missed rebase.

aaronrobertshaw replied:
No worries, when I reviewed this it was the latest on both PRs, at least on GitHub. Figured we only needed the nit on one.

aaronrobertshaw commented on packages/patterns/src/components/create-pattern-modal.js:
```diff
@@ -84,7 +99,9 @@ export default function CreatePatternModal( {
 						placeholder={ __( 'My pattern' ) }
 					/>
-
+					<CategorySelector
+						onCategorySelection={ handleCategorySelection }
+					/>
```
If we're updating the create pattern modal in the new patterns package, should we also update the reusable blocks package to maintain parity until it is officially deprecated?

glendaviesnz replied:
I think we should only backport preexisting bug fixes to the `reusable-blocks` package, not new features.

aaronrobertshaw replied:
So anywhere using the old package can only create uncategorized patterns? Do we have an easy way for users to belatedly add categories to a pattern?

glendaviesnz replied:
categories can be modified from `/wp-admin/edit.php?post_type=wp_block`

<img width="794" alt="Screenshot 2023-08-24 at 4 59 18 PM" src="https://github.com/WordPress/gutenberg/assets/3629020/24595175-db22-47bb-9eb4-2d7dcbb00c83">

and we should add that category update option to the inspector panel when editing in the site editor as well.

glendaviesnz replied:
> So anywhere using the old package can only create uncategorized patterns?

Yep, but at this point that is only unknown 3rd party plugins, and we probably shouldn't assume they want the ability to add categories added - if they do they should opt in to using the new patterns package, which we can document in a dev note. I think we should at least add some deprecation warnings to the `reusable-blocks` module as deprecated for 6.4.

aaronrobertshaw reviewed (commented):
Looking good so far @glendaviesnz ✨ 

It's testing pretty well for me. I've left some questions and comments inline.

Other than those, my main thoughts were around the display of the categories in the sidebar screen. 

1. I'd expect the `Uncategorized` pattern to display at the bottom of the categories. 
2. Might just be me, but now user patterns have merged, the separation of `My patterns` from the rest of the categories felt a bit off. I'm no designer though 😅

SaxonF commented:
Looking really good @glendaviesnz !

> Might just be me, but now user patterns have merged, the separation of My patterns from the rest of the categories felt a bit off. I'm no designer though

I think we can now change "My patterns" to "All patterns" and move to the top of the categories list (no gap in-between). It should include all patterns regardless of source. We can work on a follow up issue to filter patterns by author/source.

We should also update the copy action so it just copies into the same category. Instead of "copy", maybe we can think of a term that's similar to "remix"? 

<img width="1512" alt="image" src="https://github.com/WordPress/gutenberg/assets/1072756/99b5c547-895e-4b71-867c-b0a8abb28c47">

aaronrobertshaw commented:
> We should also update the copy action so it just copies into the same category. Instead of "copy", maybe we can think of a term that's similar to "remix"?

This originally started out as "Duplicate" and later updated to "Copy to My patterns" as the duplicate couldn't be assigned to the same theme pattern category. 

Now the duplicated pattern can reside in the same category, can we just revert back to this option being "Duplicate", like non-theme patterns? 

We can also make it such that after duplicating and navigating back to the Patterns page, it goes to the original category not "My patterns", which should remove some forked logic and simplify things.

glendaviesnz committed:
Remove hard coding of pattern type

glendaviesnz committed:
Switch to using new pattern categories map from core data store

glendaviesnz committed:
Remove my patterns section

glendaviesnz committed:
Get all-patterns category to display

glendaviesnz committed:
Make all patterns the default

glendaviesnz committed:
Add categories to pattern duplication

glendaviesnz committed:
Fix bug with category duplication

glendaviesnz commented:
@SaxonF, @aaronrobertshaw I have replaced `My patterns` with `All patterns` and also added categories into the duplication process.

It would be good to add some filters at the top of the pattern list page to just display own synced or unsynced patterns, so have left some of that code in place, but I agree with you @SaxonF that we should handle that in a follow-up issue/PR.

aaronrobertshaw committed:
Fix typos

aaronrobertshaw reviewed (commented):
Thanks for all the effort on pattern categories @glendaviesnz 🙇 

I took the liberty to push some fixes for a few typos and update the PR title to something more accurate. Hope you don't mind.

The pattern categories in the sidebar are looking good! All the counts appeared to be correct for me. 

I hit a few issues around duplicating patterns though. The first time I attempted to duplicate a theme pattern I got an error via the snackbar but not thinking clearly, I refreshed the page before checking the logs for errors. After that refresh, I couldn't replicate this issue. 

This was the snackbar notice in case it helps:

<img width="315" alt="Screenshot 2023-08-25 at 5 31 55 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/d5c858c4-c445-4b47-8e4a-8f2e9e829b5a">

After successfully duplicating a theme pattern a couple of times, I attempted to duplicate one of those copies. This resulted in an error both in the snackbar and the console showing  a 400 response from the `wp_pattern_category` endpoint.

<img width="388" alt="Screenshot 2023-08-25 at 5 33 13 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/d895f94b-2ab7-49f9-86fb-38908938f833">

<img width="941" alt="Screenshot 2023-08-25 at 5 35 00 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/c2dbad54-e4bd-4f2b-9f7c-7215669ac627">

I'm out of time today but happy to help did deeper on this Monday.

jasmussen commented:
Just testing this one, I tried creating a pattern and assigning it a non-existant category called "Newsletter":

<img width="1824" alt="Pattern Categories" src="https://github.com/WordPress/gutenberg/assets/1204802/ec72233d-f459-4d8b-baa2-46eaad86ac88">

But even upon reloading the Patterns section, I'm not seeing a new "Newsletter" category show up. Am I doing something wrong?

glendaviesnz commented:
> But even upon reloading the Patterns section, I'm not seeing a new "Newsletter" category show up. Am I doing something wrong?

Thanks for testing @jasmussen  - this is an issue with the default tag input component as it doesn't register the tag unless you hit enter or add a ','  - I already have an open task to fix this somehow on the [tracking issue](https://github.com/WordPress/gutenberg/issues/53164).

jasmussen commented:
Thanks for the clarification! This is exciting work.

glendaviesnz committed:
Fix bugs with duplicating patterns

glendaviesnz commented:
@aaronrobertshaw I think I have [fixed](https://github.com/WordPress/gutenberg/pull/53837/files/aba7fb4518ae0409a85fc3c4d549a4421c8624b2..cdb694002fdcd4cb213ef59bedef7f82bfed4ea3) the issues with duplicating patterns.

aaronrobertshaw commented on packages/edit-site/src/components/page-patterns/duplicate-menu-item.js:
```diff
@@ -119,17 +143,18 @@ export default function DuplicateMenuItem( {
 				item.title
 			);
-
+			const patternCategory = await findOrCreateTerm( categoryId );
```
This might cause two issues:
1. New patterns may be assigned to the generic "All patterns"
2. Only assigns the pattern to the current category rather than all the categories of the original pattern.

glendaviesnz replied:
I have pushed an update that uses the categories of the pattern rather than the current category listing

aaronrobertshaw reviewed (commented):
Appreciate the fixes so far @glendaviesnz 👍 

Testing this PR again, I've found a few more issues.

1. When duplicating a pattern with the "All patterns" category selected, you get a duplicate "All patterns" displayed in the sidebar (in one test this required a refresh to present itself). Duplicating a pattern in this way also doesn't put the new pattern in the same categories as the original. 
2. Duplicating some theme patterns also didn't seem to be getting the correct category assigned to them. The duplicate would show in the "All patterns" category but not the original theme pattern's category, at least until the page is refreshed.

It's probably easiest to replicate these issues after deleting all custom patterns, and the pattern category taxonomy terms.


https://github.com/WordPress/gutenberg/assets/60436221/1d9b5114-ea26-4ab7-afb8-476e18a4cb5b

glendaviesnz committed:
base the duplicated pattern categories on the actual pattern categories rather than the category of the current list

glendaviesnz commented:
> Duplicating some theme patterns also didn't seem to be getting the correct category assigned to them. 

Seems to be a caching issue somewhere, will try and work out why sometimes the category lists are not being refreshed when new patterns are added to them

glendaviesnz committed:
invalidate the pattern category cache when a new category is added

glendaviesnz commented:
@aaronrobertshaw [this change](https://github.com/WordPress/gutenberg/pull/53837/commits/d6b1c5d4ec029bab7cf159b7834c45155445dcd1) should fix the issue with theme patterns not being added to categories without a page refresh.

aaronrobertshaw reviewed (approved):
Nice work polishing up this PR @glendaviesnz 👍 

It's now testing as advertised for me. The previous issues I encountered have all been resolved.

I think this one is ready to merge into the parent PR when the time is right.

glendaviesnz committed:
Add sync filter back in for patterns

glendaviesnz commented:
This was merged with some failing tests, but  is going into a feature branch which needs rebasing to get the failing tests resolved, so failing tests will be sorted on that branch prior to merging to trunk.
